rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,152 @@
+   retrieve either the first or last stored value.
+
+## Drawbacks
+
+Metrics sinks that only support a single value per tag will need to be reworked accordingly for the
+new value type. Additionally, these sinks may change their behavior for sources that have been
+producing multiple tag values.
+
+## Prior Art
+
+The Datadog agent stores metric tags as a simple set of strings, equivalent to `HashSet<String>`. It
+does some clever hashing and deduplication internally to make this work efficiently. However, it
+doesn't do anything more interesting with the tags than adding and removing whole strings, which
+does not cover all our use cases.
+
+## Alternatives
+
+The Datadog agent represents the tags as a simple set of strings, ie `HashSet<String>` or
+`BTreeSet<String>`, where the key/value implications are just an interpretation detail for
+consumers. This is by far the simplest possible storage. However, Vector needs to be able to access
+tags by name for manipulation, making this representation more challenging. Through clever use of
+`BTreeSet::range` and a wrapper type for the tags, this could be made to work, but it is unclear if
+the benefits would be worth the additional complexity.
+
+When retaining `BTreeMap` as the top-level container, there are a number of options for the value
+that would support this feature but with different semantics:
+
+1. `Vec<String>` — Retains the ordering of tags as they appear, but allows for duplicate values and
+   cannot support both bare tags and multiple values simultaneously.
+1. `Vec<Option<String>>` — Same as above but supports bare tags and mutiple values simultaneously.
+1. `BTreeSet<Option<String>>` — Duplicate values are merged but are sorted, likely putting the bare
+   tag first for single-value uses.
+
+There are also at least two other container types that could possibly support this use case:
+
+[multimap](https://docs.rs/multimap/latest/multimap/) is a wrapper around a `HashMap` with a `Vec`
+in the value position. However, it does not sort the keys on retrieval, which changes the ordering
+guarantee that `BTreeMap` is currently providing.
+
+```rust
+type MetricTags = multimap::MultiMap<String, Option<String>>;
+```
+
+[multi_index_map](https://crates.io/crates/multi_index_map) supports both ordering of the tag names
+and multiple values. It is not clear, however, if there is any way to control for unique values for
+the same name without also preventing the same value appearing in different tags, nor how those
+values would be ordered.
+
+```rust
+#[derive(MultiIndexMap, Clone, Debug)]
+struct Tag {
+    #[multi_index(ordered_non_unique)]
+    name: String,
+    value: Option<String>,
+}
+
+type MetricTags = MultiIndexTagMap;
+```
+
+## Outstanding Questions
+
+- VRL doesn't have any way of _adding_ a tag to a metric, only replacing or deleting, nor does it
+  really have support for creating bare tags. Does it need additional functions to match this
+  support?
```

Comment:
This doesn't seem correct? https://vector.dev/docs/reference/configuration/transforms/remap/#examples-modify-metric-tags

Reply:
That was unclear wording. Here's what I mean: If a metric already has a tag `a:b`, AFAIK there is no way in VRL to set `a:c` and keep the previous value — the new value replaces the existing values for the tag `a`.

Reply:
Ah, yes - but isn't that more about the nature of the data model, rather than VRL's interaction?

Reply:
Right. The point of the above question is to ask how VRL could be made to fully support the enhanced data model.